File: arduino/connector.cpp
#include "connector.h"

void startConnector(unsigned long baudrate) {
    Serial.begin(baudrate);
}

void read_serial() {
    static bool escaping = false;
    static bool receiving = false;
    static uint8_t curr_byte = 0x00;
    
    while (Serial.available()) {
        curr_byte = Serial.read();
        // Check if received escape byte in non-escaping mode
        if (!escaping && curr_byte == ESC) { // 1
            escaping = true;
            continue;
        }
        // Seems like it's a message start
        if (!escaping && curr_byte == STX) { // 2
            receiving = true;
            msg_len = 0;
        }
        // Seems like it's a message end
        else if (receiving && !escaping && curr_byte == ETX) { // 3
            receiving = false;
            // Check if message was properly received
            if (request_valid())
                Serial.write(ACK);
            else {
                Serial.write(NAK);
                return;
            }
            
            // Execute command and place reply in msg
            execute_command();
            
            // Set length
            msg[0] = (uint8_t) (msg_len+1 >> 8);
            msg[1] = (uint8_t) (msg_len+1 & 255);
            
            // Calculate CRC
            msg[msg_len-1] = msg[0];
            for (uint16_t x=1; x<msg_len-1; x++)
                msg[msg_len-1] ^= msg[x];
                
            // Send reply and wait for ACK
            for (uint8_t x=0; x<3; x++) {
                Serial.write(STX);
                for (uint16_t b=0; b<msg_len; b++) {
                    if (msg[b] == STX || msg[b] == ETX || msg[b] == ESC)
                        Serial.write(ESC);
                    Serial.write(msg[b]);
                }
                Serial.write(ETX);
                delay(SEND_TIMEOUT);
                uint8_t reply = Serial.read();
                if (reply == ACK)
                    break;
            }
            if (planning_reset) {
                planning_reset = false;
                Commands::soft_reset();
            }
                
            return;
        }
        // Seems like it's a new message byte
        else if (receiving) { // 4
            msg[msg_len] = curr_byte;
            msg_len += 1;
        }
        if (escaping)
            escaping = false;
    }
}

bool request_valid() {
    // Check length
    uint16_t msg_extracted_len = (uint16_t) msg[0] << 8 | msg[1];
    if (msg_len+1 != msg_extracted_len)
        return false;
    // Check CRC
    uint8_t CRC = 0;
    for (uint16_t i=0; i<msg_len-1; i++)
        CRC ^= msg[i];  
    if (CRC != msg[msg_len-1])
        return false;
    // Message OK
    
    return true;
}

void execute_command() {
    uint8_t curr_command = msg[3];
    // Below, msg_len will be counted without ETX
    switch (curr_command) {
        case 0x21:  // Soft reset
            msg_len = 6;
            msg[3] = 0x00;
            planning_reset = true;
            break;
        case 0x22:  // Get buffer size
            msg_len = 8;
            msg[3] = 0x00;
            msg[5] = (uint8_t) (Commands::get_buffer_size() >> 8);
            msg[6] = (uint8_t) (Commands::get_buffer_size() & 255);
            break;
        case 0x31:  // Set pin mode
            msg_len = 6;
            msg[3] = 0x00;
            Commands::set_pin_mode(msg[4], msg[5]);
            break;
        case 0x40:  // Digital read
            msg_len = 7;
            msg[3] = 0x00;
            msg[5] = (uint8_t) Commands::digital_read(msg[4]);
            break;
        case 0x41:  // Digital write
            msg_len = 6;
            msg[3] = 0x00;
            Commands::digital_write(msg[4], (bool) msg[5]);
            break;
        case 0x42:  // Digital write timeout
            msg_len = 6;
            msg[3] = 0x00;
            Commands::digital_write_timeout(msg[4], (bool) msg[5], ((uint16_t) msg[6] << 8 | msg[7]));
            break;
        case 0x43:  // Digital write sequence
            msg_len = 6;
            msg[3] = 0x00;
            Commands::digital_write_sequence(msg[4], (bool) msg[5], msg[6], msg);
            break;
        case 0x50: { // Analog read
            msg_len = 8;
            msg[3] = 0x00;
            uint16_t value = Commands::analog_read(msg[4]);
            msg[5] = (uint8_t) (value >> 8);
            msg[6] = (uint8_t) (value & 255);
            break;
        }
        case 0x51:  // Analog write
            msg_len = 6;
            msg[3] = 0x00;
            Commands::analog_write(msg[4], msg[5]);
            break;   
    }      
    msg[4] = curr_command;
}

File: commpyduino/connector.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
__author__ = 'mFoxRU'

from typing import Iterable
from itertools import chain

from serial import Serial

from commpyduino.protocol import CommProtocol
from commpyduino.status_codes import Status


class Connector:

    def __init__(self, port: str, baudrate: int=115200):
        super().__init__()
        self._serial = Serial()
        self._serial.setPort(port)
        self._serial.baudrate = baudrate
        self._serial.timeout = .1

        self._protocol = CommProtocol()

        self._last_command = None
        self._last_reply = None

        self.commands = {
            'soft_reset': self.soft_reset,
            'get_buffer_size': self.get_buffer_size,
            'set_pin_mode': self.set_pin_mode,
            'digital_read': self.digital_read,
            'digital_write': self.digital_write,
            'digital_write_timeout': self.digital_write_timeout,
            'digital_write_sequence': self.digital_write_sequence,
            'analog_read': self.analog_read,
            'analog_write': self.analog_write,
        }

    def _communicate(self, cmd: int, data: bytearray=None) -> dict:
        self._last_command = cmd
        message = self._protocol.compose_request(cmd, data)
        if not self._serial.isOpen():
            try:
                self._serial.open()
            except Exception as e:
                self._last_reply = {'status': Status.REQ_ERR,
                                    'frame_id': self._protocol.frame_id}
                return self._last_reply

        self._serial.readall()
        # Send request
        for _try in range(3):
            self._serial.write(message)
            rep = self._serial.read(1)
            if rep[0] == self._protocol.ACK:
                break

            if _try == 2:
                self._last_reply = {'status': Status.NO_ACK,
                                    'frame_id': self._protocol.frame_id}
                return self._last_reply
        # Read reply
        for _try in range(3):
            while not self._serial.in_waiting:
                pass
            reply = self._serial.readall()
            self._last_reply = self._protocol.decompose_reply(bytearray(reply))
            # Check if reply is valid
            if self._last_reply['status'] not in (
                    Status.LEN_ERR, Status.CRC_ERR, Status.ID_ERR):
                # Seems legit
                self._serial.write(self._protocol.ACK)
                return self._last_reply

            # Seems not legit
            self._serial.write(self._protocol.NAK)

            # No more tries, deal with it
            if _try == 2:
                if 'frame_id' not in self._last_reply:
                    self._last_reply['frame_id'] = self._protocol.frame_id
                return self._last_reply

    @property
    def last_command(self) -> int:
        return self._last_command

    @property
    def last_reply(self) -> dict:
        return self._last_reply

    def soft_reset(self) -> dict:
        return self._communicate(0x21)

    def get_buffer_size(self) -> dict:
        reply = self._communicate(0x22)
        if 'data' in reply and len(reply['data']) == 2:
            reply['data'] = reply['data'][0] * 256 + reply['data'][1]
        return reply

    def set_pin_mode(self, pin: int, mode: int) -> dict:
        payload = bytearray((pin, mode))
        return self._communicate(0x31, payload)

    def digital_read(self, pin: int) -> dict:
        payload = bytearray((pin, ))
        reply = self._communicate(0x40, payload)
        if 'data' in reply and len(reply['data']) == 1:
            reply['data'] = reply['data'][0]
        return reply

    def digital_write(self, pin: int, value: int) -> dict:
        payload = bytearray((pin, value))
        return self._communicate(0x41, payload)

    def digital_write_timeout(self, pin: int, value: int, timeout: int) ->dict:
        payload = bytearray((pin, value, timeout // 256, timeout % 256))
        return self._communicate(0x42, payload)

    def digital_write_sequence(self, pin: int, value: int, n: int,
                               timeouts: Iterable) -> dict:
        payload = bytearray((pin, value, n))
        payload.extend(
            chain.from_iterable((x // 256, x % 256) for x in timeouts))
        return self._communicate(0x43, payload)

    def analog_read(self, pin: int) -> dict:
        payload = bytearray((pin, ))
        reply = self._communicate(0x50, payload)
        if 'data' in reply and len(reply['data']) == 2:
            reply['data'] = reply['data'][0] * 256 + reply['data'][1]
        return reply

    def analog_write(self, pin: int, value: int) -> dict:
        payload = bytearray((pin, value))
        return self._communicate(0x51, payload)
